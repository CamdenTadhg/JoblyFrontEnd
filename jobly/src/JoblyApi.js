import axios from "axios";

// const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";
const BASE_URL = 'http://localhost:3001';

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

class JoblyApi {
  // the token for interaction with the API will be stored here.
  static token;

  static async request(endpoint, data = {}, method = "get") {
    console.log("API Call:", endpoint, data, method);

    //there are multiple ways to pass an authorization token, this is how you pass it in the header.
    //this has been provided to show you another way to pass the token. you are only expected to read this code for this project.
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { authorization: `Bearer ${JoblyApi.token}` };
    const params = (method === "get")
        ? data
        : {};

    try {
      console.log('headers are', headers);
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    }
  }

  // Individual API routes
  /** Login a user and receive back a token. LoginData = {username, password}*/
  static async login(loginData){
    try {
      let res = await this.request(`auth/token`, loginData, 'post');
      console.log('res.token is ', res.token);
      JoblyApi.token=res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Signup a user and receive back a token. signupData = {username, password, firstName, lastName, email} */
  static async signup(signupData){
    try{
      let res = await this.request(`auth/register`, signupData, "post");
      JoblyApi.token = res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Get list of all companies */
  static async getAllCompanies(){
    let res = await this.request(`companies`);
    return res.companies;
  }

  /** Get a filtered list of companies. searchFilters = {minEmployees, maxEmployees, name} */
  static async getFilteredCompanies(searchFilters) {
    let {minEmployees, maxEmployees, name} = searchFilters;
    let url = 'companies?';
    if (minEmployees) {
      url = url + `minEmployees=${minEmployees}&`;
    }
    if (maxEmployees){
      url = url + `maxEmployees=${maxEmployees}&`;
    }
    if (name) {
      url = url + `name=${name}`
    }
    let res = await this.request(url);
    console.log(res);
    return res.companies;
  }

  /** Get details on a company by handle. */
  static async getCompany(handle) {
    let res = await this.request(`companies/${handle}`);
    console.log(res.company);
    return res.company;
  }

  /**Get list of all jobs */
  static async getAllJobs(){
    let res = await this.request(`jobs`)
    return res.jobs;
  }

  /** Get a filtered list of jobs. searchFilters = {salary, equity, title}*/
  static async getFilteredJobs(searchFilters){
    let {salary, equity, title} = searchFilters;
    let url = 'jobs?'
    if (salary){
      url = url + `minSalary=${salary}&`;
    }
    if (equity) {
      url=url+`hasEquity=${equity}&`;
    }
    if(title) {
      url= url + `title=${title}`;
    }
    let res = await this.request(url);
    return res.jobs;
  }

  /** Get a user's details */
  static async getUserDetails(username){
    let res = await this.request(`users/${username}`);
    return res.user;
  }

  /** Edit a user's details. userData = {firstName, lastName, password, email} */
  static async editUserDetails(username, userData){
    let res = await this.request(`users/${username}`, userData, 'patch');
    console.log('res is', res);
    return res.user;
  }

  /** Delete a user's account */
  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, 'delete');
    return res.deleted;
  }

  /**Create job application for a user */
  static async applyToJob(username, jobId){
    let res = await this.request(`users/${username}/jobs/${jobId}`, {}, "post");
    return res;
  }
}

export default JoblyApi;
