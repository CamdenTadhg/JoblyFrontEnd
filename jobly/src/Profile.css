.Profile{
    position: absolute;
    top: 65px;
    left: 50%;
    transform: translate(-50%);
    margin: 10px;
    padding: 10px;
    border-radius: 5px;
    background-color: white;
    padding: 15px;
    width: 25vw;
}

.Profile label{
    margin: 5px;
}

.Profile input{
    margin: 0px 3px;
    border-radius: 5px;
    border: 1px solid black;
}

.input-item{
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.Profile-text-loading {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

.disabled{
    background-color: lightgrey;
}

.Profile-jobs{
    position: absolute;
    left: 50%;
    transform: translate(-50%);
    top: 350px
}