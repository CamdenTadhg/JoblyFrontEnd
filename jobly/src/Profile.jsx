import React, {useState} from 'react';
import {useNavigate} from 'react-router-dom';
import JobCard from './JobCard';
import './Profile.css';

function Profile(currentUser){
    const [formData, setFormData] = useState(currentUser);
    const [error, setError] = useState('');
    const navigate = useNavigate();

    const handleChange = (event) => {
        const {name, value} = event.target;
        setFormData(fData => ({
            ...fData,
            [name]: value
        }));
    }


    const handleSubmit = async (event) => {
        event.preventDefault();
        const {firstName, lastName, email} = formData;
        const editData = {firstName, lastName, email};
        try {
            await editProfile(editData);
        } catch (error){
            setError(error);
        }
    }

    return(
        <div className='Profile'>
            <form onSubmit={handleSubmit}>
                <h5>User Profile</h5>
                <div className="input-item">
                    <label htmlFor='username'>Username: </label>
                    <input type='text' id='username' name='username' value={FormData.username} disabled required/>
                </div>
                <div className='input-item'>
                    <label htmlFor='firstName'>First Name:</label>
                    <input type='text' id='firstName' name='firstName' value={FormData.firstName} onChange={handleChange} required/>
                </div>
                <div className='input-item'>
                    <label htmlFor='lastName'>Last Name:</label>
                    <input type='text' id='lastName' name='lastName' value={FormData.lastName} onChange={handleChange} required/>
                </div>
                <div className='input-item'>
                    <label htmlFor='email'>Email:</label>
                    <input type='text' id='email' name='email' value={FormData.email} onChange={handleChange} required/>
                </div>
            </form>
        </div>
    )
}

export default Profile;