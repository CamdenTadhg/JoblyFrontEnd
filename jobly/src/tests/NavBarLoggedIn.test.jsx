import {fireEvent, render, waitFor, cleanup} from '@testing-library/react';
import {test, expect, vi, afterEach, beforeEach} from 'vitest';
import {MemoryRouter} from 'react-router-dom';
import Profile from '../Profile';
import JoblyApi from '../JoblyApi';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import UserContext from '../contexts/userContext'

const Wrapper = ({children}) => {
    return (
        <MemoryRouter>
            <UserContext.Provider value='testuser'>
                {children}
            </UserContext.Provider>
        </MemoryRouter>
    );
};

afterEach(cleanup);

test('matches the snapshot', () => {
    const navbar = render(<NavBar/>, {wrapper: Wrapper});
    expect(navbar).toMatchSnapshot();
});

test('displays the correct content', () => {
    const {getByText, queryByText} = render(<NavBar/>, {wrapper: Wrapper});
    expect(getByText('Jobs')).toBeInTheDocument();
    expect(queryByText('Login')).not.toBeInTheDocument();
});


test('logs a user out', () => {})