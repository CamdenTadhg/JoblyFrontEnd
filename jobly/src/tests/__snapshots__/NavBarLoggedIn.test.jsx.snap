// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav>
        <nav
          class="fixed-top NavBar navbar"
        >
          <div
            class="container-fluid"
          >
            <a
              aria-current="page"
              class="navbar-brand active"
              href="/"
            >
              Jobly
            </a>
            <ul
              class="navbar-center nav justify-content-center"
            >
              <a
                class="nav-item"
                href="/companies"
              >
                Companies
              </a>
              <a
                class="nav-item"
                href="/jobs"
              >
                Jobs
              </a>
            </ul>
            <ul
              class="nav"
            >
              <a
                class=""
                href="/profile"
              >
                Profile
              </a>
              <a
                aria-current="page"
                class="active"
                href="/"
              >
                Logout 
                testuser
              </a>
            </ul>
          </div>
        </nav>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav>
      <nav
        class="fixed-top NavBar navbar"
      >
        <div
          class="container-fluid"
        >
          <a
            aria-current="page"
            class="navbar-brand active"
            href="/"
          >
            Jobly
          </a>
          <ul
            class="navbar-center nav justify-content-center"
          >
            <a
              class="nav-item"
              href="/companies"
            >
              Companies
            </a>
            <a
              class="nav-item"
              href="/jobs"
            >
              Jobs
            </a>
          </ul>
          <ul
            class="nav"
          >
            <a
              class=""
              href="/profile"
            >
              Profile
            </a>
            <a
              aria-current="page"
              class="active"
              href="/"
            >
              Logout 
              testuser
            </a>
          </ul>
        </div>
      </nav>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
